oci/layout/oci_archive_src.go

```diff
@@ -0,0 +1,81 @@
+package layout
+
+import (
+	"io"
+
+	"os"
+
+	"github.com/containers/image/types"
+	"github.com/containers/storage/pkg/archive"
+	digest "github.com/opencontainers/go-digest"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+var ociImgSrc types.ImageSource
+
+type ociArchiveImageSource struct {
+	ref        ociArchiveReference
+	descriptor imgspecv1.Descriptor
+}
+
+// newImageSource untars the tar file to a directory and
+// returns an ImageSource for reading from an existing directory.
+func newImageSourceArchive(ref ociArchiveReference) (types.ImageSource, error) {
+	path := ociRef.(ociReference).resolvedDir
+	if err := unTar(path); err != nil {
+		return nil, errors.Wrapf(err, "error untaring %q", path)
+	}
+	ociImgSrc, _ = newImageSource(ociRef.(ociReference))
+	return &ociArchiveImageSource{ref: ref, descriptor: ociImgSrc.(*ociImageSource).descriptor}, nil
+}
+
+// Reference returns the reference used to set up this source.
+func (s *ociArchiveImageSource) Reference() types.ImageReference {
+	return ociImgSrc.Reference()
+}
+
+// Close removes resources associated with an initialized ImageSource, if any.
+// Close tars the untarred directory after the image has been retrieved from source
+func (s *ociArchiveImageSource) Close() error {
+	if err := ociImgSrc.Close(); err != nil {
+		return err
+	}
+	path := ociRef.(ociReference).resolvedDir
+	if err := tar(path); err != nil {
```

Comment:
`ociArchiveImageSource` really must not modify the original file in any way, not even with the best intentions of replacing it later:

1. In the meantime others expecting to use the original file at that path can't do so.
1. The process may crash, leaving the original replaced by an unexpected directory forever.
1. Nothing guarantees that untar+tar will reproduce the original byte-for-byte exactly. 

Similarly to the discussion about `ociArchiveImageDestination`, `newImageSourceArchive` should create a temporary directory and extract the image there.

(Alternatively, it would be also possible to implement `ociArchiveImageSource` which does not need to extract the tarball at all, by just seeking around the original tarball, like `docker/tarfile/Source.openTarComponent` and callers do.  Then `oci/archive` and `oci/layout` could share the format implementation, differing only in the backend used to access file components. But that would only work with seekable inputs; extracting the tarball and calling the `layout` implementation directly is perfectly fine.)